uqcsbot/whatweekisit.py

```diff
@@ -0,0 +1,160 @@
+        )
+
+        semesters.append(
+            Semester(semester_title.text, start_date, end_date, week_names)
+        )
+
+    return semesters
+
+
+def get_semester_week(
+    semesters: List[Semester], checked_date: datetime
+) -> Tuple[Optional[str], Optional[str], Optional[str]]:
+    """
+    Gets the name of the semester, week and weekday of the date that's to be checked
+        Parameters:
+            checked_date: the given date that we'll be checking the semester and week for
+        Returns:
+            A tuple containing the semester, the name of the week we're in, and the weekday
+            respectively. 
+    """
+    semester_name = None
+    week_name = None
+    weekday = None
+
+    for semester in semesters:
+        # Check if the current date is within the semester range
+        if semester.start_date <= checked_date <= semester.end_date:
+            # If it is, we figure out what week it is
+            days_since_start = (checked_date - semester.start_date).days
+            week_index = days_since_start // 7
+
+            weekday = checked_date.strftime("%A")
+            week_name = semester.weeks[week_index]
+            if "week" not in week_name.lower():
+                # Accounts for & makes things like "Revision", "Exam", "Pause" a bit nicer
+                week_name = week_name + " Week"
+            semester_name = semester.name
+
+            break
+
+    return semester_name, week_name, weekday
+
+
+class WhatWeekIsIt(commands.Cog):
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+
+    @commands.command()
+    @loading_status
+    async def whatweekisit(self, ctx: commands.Context, *args):
+        """
+        `!whatweekisit [SPECIFIED_DATE]` - Sends information about which
+        semester, week and weekday it is on the specified date (in %d/%m/%Y) --
+        if there's no specified date, it takes it to be the current one.
+        """
+        if len(args) > 1:
+            await ctx.send("No more than one argument (specified date) is required")
+            return
+        elif len(args) == 1:
+            try:
+                check_date = string_to_date(args[0])
+            except ValueError:
+                await ctx.send(f"Specified date should be in format `{DATE_FORMAT}`")
+                return
+        else:
+            check_date = datetime.now()
+
+        calendar_page = requests.get(MARKUP_CALENDAR_URL)
+        if (calendar_page.status_code != requests.codes.ok):
+            await ctx.send("An error occurred, please try again.")
+
+        semesters = get_semester_times(calendar_page.text)
+
+        semester_name, week_name, weekday = get_semester_week(semesters, check_date)
+        if not semester_name:
+            date = date_to_string(check_date)
+            message = f"University isn't in session on {date}, enjoy the break :)"
+        else:
+            message = "The week we're in is:\n"
```

Comment:
super nit: this shows even when a manual date is given, so it's not always “the week we're in".